rclpy/rclpy/node.py

```diff
@@ -233,41 +318,164 @@ def get_parameter(self, name: str) -> Parameter:
+        If the callback doesn't succeed for a given parameter, it won't be set and either an
+        unsuccessful result will be returned for that parameter, or an exception will be raised
+        according to `raise_on_failure` flag.
+
+        :param parameter_list: List of parameters to set.
+        :param setter_func: Method to set the parameters. Use :func:`_set_parameters_atomically`
+            to declare and set the parameters without checking if they had been declared,
+            or :func:`set_parameters_atomically` to check if they had been declared beforehand.
+        :return: The result for each set action as a list.
+        :raises: InvalidParameterValueException if the user-defined callback rejects the
+            parameter value and raise_on_failure flag is True.
         """
+        if setter_func is None:
+            setter_func = self._set_parameters_atomically
+
         results = []
         for param in parameter_list:
-            if not isinstance(param, Parameter):
-                raise TypeError("parameter must be instance of type '{}'".format(repr(Parameter)))
-            results.append(self.set_parameters_atomically([param]))
+            result = setter_func([param])
+            if raise_on_failure and not result.successful:
+                raise InvalidParameterValueException(param.name, param.value)
+            results.append(result)
         return results
 
     def set_parameters_atomically(self, parameter_list: List[Parameter]) -> SetParametersResult:
         """
-        Atomically set parameters for the node.
+        Set the given parameters, all at one time, and then aggregate result.
+
+        If any parameter in the list was not declared beforehand and undeclared parameters are not
+        allowed for the node, this method will raise a ParameterNotDeclaredException exception.
+
+        Parameters are set all at once. If setting a parameter fails due to not being declared,
+        then no parameter will be set set. Either all of the parameters are set or none of them
+        are set.
+
+        If undeclared parameters are allowed, then all the parameters will be implicitly
+        declared before being set even if they were not declared beforehand.
+
+        If a callback was registered previously with :func:`set_parameters_callback`, it will be
+        called prior to setting the parameters for the node only once for all parameters.
+        If the callback prevents the parameters from being set, then it will be reflected in the
+        returned result; no exceptions will be raised in this case.
+        For each successfully set parameter, a :class:`ParameterEvent` message is
+        published.
+
+        If the value type of the parameter is NOT_SET, and the existing parameter type is
+        something else, then the parameter will be implicitly undeclared.
+
+        :param parameter_list: The list of parameters to set.
```

Comment:
I see a problem here (I think). If I declare a parameter as read_only or with a range as part of the descriptor, and then I set it without a descriptor, doesn't the descriptor default to `ParameterDescriptor()` and therefore overwrite the descriptor that I set when declaring?

Reply:
- Attempting to set a `read_only` parameter should fail with an exception. That was not implemented here, and was not implemented in `rclcpp` yet AFAIK. If I'm not mistaken, that should be taken into account in a follow-up PR.
- Yes, the descriptor would override the existing one; I'm not sure if that would be expected behavior or not (it could be as long as the existing descriptor is not `read_only`?).
I don't think it's a good idea to change `set_parameters` signature at this point to take just `name` and `value`. On the other hand, the `Parameter` class as of `master` branch only supports the default `ParameterDescriptor`, so I think it's a good idea to extend it as in this PR to support custom ones. A possible compromise solution would be to ignore the `ParameterDescriptor` sent to this method, and use the existing one instead if it exists if we don't want to override the descriptor.

Reply:
Setting a read only parameter should fail in rclcpp, but that might be an oversight right now, there is a test for undeclare:

https://github.com/ros2/rclcpp/blob/7ed130cf7ac7b08158d6aa2e8bd808e0cccee215/rclcpp/test/test_node.cpp#L577-L588

It will fail to set it:

https://github.com/ros2/rclcpp/blob/7ed130cf7ac7b08158d6aa2e8bd808e0cccee215/rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp#L392-L397

I don't think it should be an exception necessarily, but maybe.

---

To match C++, we'd need to remove the descriptor from the Parameter class. I think that makes sense, because you want to be able to conveniently create a Parameter as a user in order to set one, e.g. `node.set_parameter(Parameter(name, value))`, and not worry about the description.

I don't know if we should let set change the description or not, in C++ it is not possible, you much undeclare and re-declare to change the description.

Personally I'd like to see something like `node.declare_parameter(Parameter, ParameterDescriptor)`, `node.set_parameter(Parameter)`, and `node.set_parameter_descriptor(ParameterDescriptor)` (the `ParameterDescriptor` contains the name already I believe. We could later add a "set parameter descriptor" function in c++ too.

Reply:
Thanks for the feedback; that makes sense.
I was confused by the fact that `Parameter` held a `ParameterDescriptor`; removing it would iron out the confusion and all the intermediate problems that arise from it. Let's go that way; I'll make those changes.

Reply:
> Personally I'd like to see something like node.declare_parameter(Parameter, ParameterDescriptor)

That could work for `declare_parameter` but not for `declare_parameters` as we discussed here: https://github.com/ros2/rclpy/pull/325#discussion_r280233234. For that I'd stick with the proposal that we have now (i.e. using separate `name`, `value`, `descriptor`).

Reply:
Sounds good to me.